### ISSUE TITLE
complete testing of getlens.f

### ISSUE THREAD

**Issue by edwardhartnett** [NOAA-EMC/NCEPLIBS-bufr#351]
The code says:

```
C> This subroutine reads the lengths of all of the individual
C> sections of a given BUFR message, up to a specified point in
C> the message.

```

So it seems like a good idea to use one of the test input files, and make sure this function comes up with the correct answer for the test file.

**Comment by AlexanderRichert-NOAA** [comment]
@edwardhartnett @jbathegit 
The remaining untested chunk of code runs when `im8b=.true.` In, e.g., intest1.F90, `setim8b(.true.)` happens with `#ifdef KIND_8`. The first problem is that this setting doesn't seem to be respected-- inside the `getlens` subroutine, `im8b=.false.` even though `setim8b` was successfully called in intest1.F90. The second problem is that if I explicitly set `im8b=.true.` in `getlens.f` and build and run `intest1_8`, I get a segfault. Would it be fair to guess that at least the latter issue relates to the removal of _8 and _d? More broadly, I'm not sure I follow why it is that _4, _8, and _d are all included in testing, but the _8 and _d libraries aren't actually buildable as-is.

**Comment by jbathegit** [comment]
Hi Alex.  We recently simplified down to just doing one build of the library (as _4), but there are still many application codes out there which link to the library and are compiled as _8 and _d.  So that's why we have the new setim8b routine (so that _8 applications can still link to this one build of the library), and it's also the reason that we still include _4, _8, and _d variants when we run our test codes (to make sure everything still works for all variants of application codes with this one build of the library).  Hope that makes sense.

As for what's going on in getlens, once the logic enters the `IF(IM8B)` block, it temporarily resets IM8B to .false., then copies all 8-byte integer arguments into 4-byte values, then recursively calls getlens again as though it were being called from a 4-byte application code.  That's why we temporarily reset IM8B to .false., so that we don't get stuck in an infinite loop.  Then, once control returns from the recursive call to getlens, it remaps all output integers from 4-bytes to 8-bytes, then resets IM8B to its original value of .true., then returns control to the calling program.

A similar `IF(IM8B)` wrapper is in place for all of the library routines which are user-callable and have integer arguments.  Another nice thing about this approach is that for any internal calls to other library subroutines inside of getlens (e.g. the internal calls to iupb and iupbs01), those routines don't now have to also temporarily set and reset im8b, because that's already been done inside of the parent getlens routine, and those routines are therefore already now being called with all 4-byte integer arguments.

As for your second problem/question, you should never be explicitly setting im8b=.true. in your test code or in any application code.  Instead, you should call subroutine setim8b to do that for you.  That's why you're getting a segfault, because it's trying to overwrite your im8b value which is a logical constant, whereas if you call subroutine setim8b then it uses an internal variable which can be internally toggled on and off as described above.

Again, I hope this is all (somewhat ;-) clear.  At any rate, for adding testing of getlens, it might be easiest for you to just add a `call getlens` line to one of the existing test codes such as intest1.F90 or intest4.F90 where crbmg is already being called, and therefore a BUFR message is already being stored inside of an integer array ibfmg which you can just pass in as the first argument to getlens and then check the output values.  Otherwise, you'll have to cook up your own similar logic to read a BUFR message from your own input BUFR file into an integer array before you can call getlens yourself.   Furthermore, if you use one of the existing test codes, then the call to setim8b is already being done for you whenever KIND=8, and you can use this overall design as a sort-of prototype for developing other new test codes of your own in the future.

**Comment by AlexanderRichert-NOAA** [comment]
To be clear, what I said about `im8b=.false.` applies before it reaches that block-- I can see that it gets unset in there, but the value set in the main program of `intest1` seems to be getting lost when it reaches the `getlens` subroutine. Maybe there's a simpler way to ask the question-- why isn't that block of code already getting hit by the _8 tests?

**Comment by jbathegit** [comment]
Sorry if my response wasn't clear enough Alex.  In this case, what's happening is that getlens isn't being called directly by intest1, but rather indirectly from within other routines that are called directly from intest1, such as upds3 and iupbs3.  So it's being called, but as noted in the second and third paragraphs of my initial reply, control is only ever going to go into the `IF(IM8B)` block in the initial routine called directly from intest1 (so, in this case, upds3 or iupbs3), and for any downstream internal calls to getlens and other routines from within that initial routine, it will never go into that block because everything has now already been copied and remapped to 4-byte integers.

So bottom line, the only way to get it to exercise the `IF(IM8B)` block inside of getlens is to call getlens directly from an application program such as the intest1 code itself.  In other words, you could just add your own additional direct call to getlens inside of intest1, and that would do the trick.  As noted in the third paragraph, it would then go into the `IF(IM8B)` block in getlens, but then not have to do so again for iupb or iupbs01 when those are called internally within getlens.

I hope this helps!

**Comment by AlexanderRichert-NOAA** [comment]
Okay, I see what I missed now. Thanks for the clarification.
